File: module2/Dequeue/Dequeue.py
class Dequeue:
    def __init__(self, capacity):
        self.items = [None] * capacity
        self.capacity = capacity
        self.size = 0
        self.left = 0
        self.right = capacity - 1

    def isEmpty(self):
        return self.size == 0
    
    def isFull(self):
        return self.size == self.capacity
    
    def enqueueLeft(self, item):
        if self.isFull():
            return False
        self.items[self.left] = item
        self.left = (self.left - 1 + self.capacity) % self.capacity
        self.size += 1
        return True
    
    def enqueueRight(self, item):
        if self.isFull():
            return False
        self.items[self.right] = item
        self.right = (self.right + 1) % self.capacity
        self.size += 1
        return True
    
    def dequeueLeft(self):
        if self.isEmpty():
            return None
        item = self.items[self.left]
        self.left = (self.left + 1) % self.capacity
        self.size -= 1
        return item

    def dequeueRight(self):
        if self.isEmpty():
            return None
        item = self.items[self.right]
        self.right = (self.right - 1 + self.capacity) % self.capacity
        self.size -= 1
        return item
    
    def peakLeft(self):
        if self.isEmpty():
            return None
        return self.items[self.left]
    
    def peakRight(self):
        if self.isEmpty():
            return None
        return self.items[self.right]
    
if __name__ == "__main__":
    capacity = 5
    dequeue = Dequeue(capacity)
        
    for i in range(3, 8):
        dequeue.enqueueRight(i)

    # [7] [6] [5] [4] [3]
    # right is at index 4 after finish enqueueRight
        
    while not dequeue.isEmpty():
        print(dequeue.dequeueRight(), end=" ")

File: module2/Dequeue/Dequeue.java
package module2.Dequeue;
public class Dequeue<T> {
    private int size, left, right, capacity;
    private T[] items;

    public Dequeue(int capacity) {
        items = (T[]) new Object[capacity];
        this.capacity = capacity;
        this.size = 0;
        this.left = 0;
        this.right = capacity - 1;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public void enqueueLeft(T item) {
        if (isFull()) {
            throw new RuntimeException("Dequeue is full");
        }
        items[left] = item;
        left = (left - 1 + capacity) % capacity;
        size++;
    }

    public void enqueueRight(T item) {
        if (isFull()) {
            throw new RuntimeException("Dequeue is full");
        }
        items[right] = item;

        // plus capacity to avoid negative index
        right = (right + 1) % capacity;
        size++;
    }

    public T dequeueLeft() {
        if (isEmpty()) {
            throw new RuntimeException("Dequeue is empty");
        }
        T item = items[left];
        left = (left + 1) % capacity;
        size--;
        return item;
    }

    public T dequeueRight() {
        if (isEmpty()) {
            throw new RuntimeException("Dequeue is empty");
        }
        T item = items[right];
        right = (right - 1 + capacity) % capacity;
        size--;
        return item;
    }

    public T peakLeft() {
        if (isEmpty()) {
            throw new RuntimeException("Dequeue is empty");
        }
        return items[left];
    }

    public T peakRight() {
        if (isEmpty()) {
            throw new RuntimeException("Dequeue is empty");
        }
        return items[right];
    }

    public static void main(String[] args) {
        int capacity = 5;
        Dequeue<Integer> dequeue = new Dequeue<>(capacity);
        
        for (int i = 3; i <= 7; i++) {
            dequeue.enqueueRight(i);
        }

        // [7] [6] [5] [4] [3]
        // right is at index 4 after finish enqueueRight
        
        while (!dequeue.isEmpty()) {
            System.out.print(dequeue.dequeueRight() + " ");
        }
    }
}
